async_producer_test.go

```diff
@@ -638,6 +639,71 @@ func TestAsyncProducerMultipleRetriesWithBackoffFunc(t *testing.T) {
 	}
 }
 
+func TestAsyncProducerWithExponentialBackoffDurations(t *testing.T) {
+	var backoffDurations []time.Duration
+	var mu sync.Mutex
+
+	topic := "my_topic"
+	maxBackoff := 2 * time.Second
+	config := NewTestConfig()
+
+	initBackoff := 100 * time.Millisecond
+	innerBackoffFunc := NewExponentialBackoff(initBackoff, maxBackoff)
+	backoffFunc := func(retries, maxRetries int) time.Duration {
+		duration := innerBackoffFunc(retries, maxRetries)
+		mu.Lock()
+		backoffDurations = append(backoffDurations, duration)
+		mu.Unlock()
+		return duration
+	}
+
+	config.Producer.Flush.Messages = 5
+	config.Producer.Return.Successes = true
+	config.Producer.Retry.Max = 3
+	config.Producer.Retry.BackoffFunc = backoffFunc
+
+	broker := NewMockBroker(t, 1)
+
+	metadataResponse := new(MetadataResponse)
+	metadataResponse.AddBroker(broker.Addr(), broker.BrokerID())
+	metadataResponse.AddTopicPartition(topic, 0, broker.BrokerID(), nil, nil, nil, ErrNoError)
+	broker.Returns(metadataResponse)
+
+	producer, err := NewAsyncProducer([]string{broker.Addr()}, config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	failResponse := new(ProduceResponse)
+	failResponse.AddTopicPartition(topic, 0, ErrNotLeaderForPartition)
+	successResponse := new(ProduceResponse)
+	successResponse.AddTopicPartition(topic, 0, ErrNoError)
+
+	broker.Returns(failResponse)
+	broker.Returns(metadataResponse)
+	broker.Returns(failResponse)
+	broker.Returns(metadataResponse)
+	broker.Returns(successResponse)
+
+	for i := 0; i < 5; i++ {
+		producer.Input() <- &ProducerMessage{Topic: topic, Value: StringEncoder("test")}
+	}
+
+	expectResults(t, producer, 5, 0)
+	closeProducer(t, producer)
+	broker.Close()
+
+	assert.Greater(t, backoffDurations[0], time.Duration(0), "Expected first backoff duration to be greater than 0")
+
+	for i := 1; i < len(backoffDurations); i++ {
+		assert.Greater(t, backoffDurations[i], time.Duration(0), "Expected backoff[%d] to be greater than 0", i)
+		assert.GreaterOrEqual(t, backoffDurations[i], backoffDurations[i-1],
+			"Expected backoff[%d] >= backoff[%d], but got %v < %v", i, i-1, backoffDurations[i], backoffDurations[i-1])
```

Comment:
Consider what your test error reports will look like, forcing this into a failure state t o check the message:
```
--- FAIL: TestAsyncProducerWithExponentialBackoffDurations (0.33s)
    async_producer_test.go:700:
                Error Trace:    /home/puellanivis/tmp/github.com/wanwenli/sarama/async_producer_test.go:700
                Error:          "115.755034ms" is not greater than or equal to "213.387835ms"
                Test:           TestAsyncProducerWithExponentialBackoffDurations
                Messages:       Expected backoff[1] >= backoff[0], but got 115.755034ms < 213.387835ms
```

You’ll notice a lot of information repeated.

Reply:
True. So I've removed the messages, except for the message in the first `assert` statement.